File: app/src/main/java/shts/jp/android/nogifeed/fragments/NewsListFragment.java
package shts.jp.android.nogifeed.fragments;

import android.os.Bundle;
import android.support.design.widget.CoordinatorLayout;
import android.support.design.widget.Snackbar;
import android.support.v4.app.Fragment;
import android.support.v4.widget.SwipeRefreshLayout;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ListView;

import com.squareup.otto.Subscribe;

import shts.jp.android.nogifeed.R;
import shts.jp.android.nogifeed.activities.NewsBrowseActivity;
import shts.jp.android.nogifeed.adapters.NewsListAdapter;
import shts.jp.android.nogifeed.api.AsyncNewsClient;
import shts.jp.android.nogifeed.entities.News;
import shts.jp.android.nogifeed.models.eventbus.BusHolder;
import shts.jp.android.nogifeed.views.dialogs.NewsTypeFilterDialog;

public class NewsListFragment extends Fragment {

    private static final String TAG = NewsListFragment.class.getSimpleName();

    @Override
    public void onResume() {
        super.onResume();
        BusHolder.get().register(this);
    }

    @Override
    public void onPause() {
        super.onPause();
        BusHolder.get().unregister(this);
    }

    private ListView listView;
    private CoordinatorLayout coordinatorLayout;
    private SwipeRefreshLayout swipeRefreshLayout;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_news_feed_list, null);
        view.findViewById(R.id.fab).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                NewsTypeFilterDialog dialog = new NewsTypeFilterDialog();
                dialog.setCallbacks(new NewsTypeFilterDialog.Callbacks() {
                    @Override
                    public void onClickPositiveButton() {
                    }
                    @Override
                    public void onClickNegativeButton() {}
                });
                dialog.show(getFragmentManager(), NewsTypeFilterDialog.class.getSimpleName());
            }
        });
        swipeRefreshLayout = (SwipeRefreshLayout) view.findViewById(R.id.refresh);
        swipeRefreshLayout.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                getAllNews();
            }
        });
        swipeRefreshLayout.setColorSchemeResources(
                R.color.primary, R.color.primary, R.color.primary, R.color.primary);

        listView = (ListView) view.findViewById(R.id.news_feed_list);
        coordinatorLayout = (CoordinatorLayout) view.findViewById(R.id.coordinator);

        getAllNews();
        return view;
    }

    private void getAllNews() {
        if (!AsyncNewsClient.get(getActivity())) {
            Snackbar.make(coordinatorLayout, R.string.failed_to_get_news, Snackbar.LENGTH_LONG)
                    .show();
            swipeRefreshLayout.setRefreshing(false);
        }
    }

    @Subscribe
    public void onGotNewsFeedList(AsyncNewsClient.GetNewsFeedCallback callback) {
        swipeRefreshLayout.setRefreshing(false);
        if (callback == null || callback.hasError()) {
            Snackbar.make(coordinatorLayout, R.string.failed_to_get_news, Snackbar.LENGTH_LONG)
                    .show();
            return;
        }
        listView.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                News news = (News) listView.getItemAtPosition(position);
                startActivity(NewsBrowseActivity.getStartIntent(getActivity(), news));
            }
        });
        listView.setAdapter(new NewsListAdapter(getActivity(), callback.newsList));
    }
}

File: app/src/main/java/shts/jp/android/nogifeed/fragments/NewsListFragment.kt
package shts.jp.android.nogifeed.fragments

import android.app.AlertDialog
import android.os.Bundle
import android.support.design.widget.Snackbar
import android.support.v4.app.Fragment
import android.support.v7.widget.DividerItemDecoration
import android.support.v7.widget.DividerItemDecoration.VERTICAL
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import kotlinx.android.synthetic.main.fragment_news_feed_list.*
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers
import rx.subscriptions.CompositeSubscription
import shts.jp.android.nogifeed.R
import shts.jp.android.nogifeed.activities.NewsBrowseActivity
import shts.jp.android.nogifeed.adapters.NewsAdapter
import shts.jp.android.nogifeed.adapters.OnNewsClickListener
import shts.jp.android.nogifeed.api.AsyncNewsClient
import shts.jp.android.nogifeed.entities.News

class NewsListFragment : Fragment() {

    companion object {
        val newInstance: NewsListFragment =
                NewsListFragment()
    }

    private val subscriptions: CompositeSubscription = CompositeSubscription()

    val adapter: NewsAdapter = NewsAdapter().apply {
        clickListener = object : OnNewsClickListener {
            override fun onClickNews(news: News) {
                startActivity(NewsBrowseActivity.getStartIntent(activity, news))
            }
        }
    }

    override fun onCreateView(inflater: LayoutInflater,
                              container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_news_feed_list, null)
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        newsFeedList.adapter = adapter
        newsFeedList.addItemDecoration(DividerItemDecoration(context, VERTICAL))
        fab.setOnClickListener { showFilterDialog() }
        refreshLayout.setOnRefreshListener { getAllNews() }
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        getAllNews()
    }

    override fun onDestroyView() {
        subscriptions.unsubscribe()
        super.onDestroyView()
    }

    private fun getAllNews() {
        subscriptions.add(AsyncNewsClient.get(activity)
                .doOnSubscribe { refreshLayout.isRefreshing = true }
                .subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ newsList ->
                    refreshLayout.isRefreshing = false
                    newsList?.let {
                        adapter.add(it)
                    } ?: run {
                        Snackbar.make(coordinatorLayout, R.string.failed_to_get_news, Snackbar.LENGTH_LONG).show()
                    }
                }, { e ->
                    refreshLayout.isRefreshing = false
                    e.printStackTrace()
                }))
    }

    private fun showFilterDialog() {
        val typeList = News.Type.getTypeList(activity)
        val filter = News.Type.getFilter(activity)

        AlertDialog.Builder(activity)
                .setTitle(R.string.news_filter_dialog_title)
                .setMultiChoiceItems(typeList, filter, { _, which, isChecked -> filter[which] = isChecked })
                .setPositiveButton(R.string.news_filter_dialog_ok, { _, _ ->
                    News.Type.setFilter(activity, filter)
                    getAllNews()
                })
                .setNegativeButton(R.string.news_filter_dialog_cancel, null).show()
    }
}
